### DISCUSSION TITLE
Request a firmware build - 2.8.0

### DISCUSSION THREAD

**Discussion by pfeerick** [EdgeTX/edgetx#2362] [General]
Ask here if you want/need specific options that are not included in the stock builds. Builds will be uploaded to [here (same link as 'By request' below)](http://files.portabledev.xyz/d/e9c2cc06ff22454e89fb/), so check it out before making a request just in case there is already a custom build with your chosen options...

The Language builds are automated builds based off of 2.8.0

*Please remember to specify your transmitter when requesting a build* ;)

- [Language builds](https://github.com/pfeerick/lang-firmwares/releases/tag/release) - zip file per transmitter containing all languages built
- [Language builds - heli](https://github.com/pfeerick/lang-firmwares/releases/tag/latest) - zip file per transmitter containing all languages built. Is still a work in progress so contains heli AND "normal" firmwares. 
- Builds for custom lua scripts is next on the agenda

- [By request](http://files.portabledev.xyz/d/e9c2cc06ff22454e89fb/)

**Comment by RipperGOT** [comment]
I leave this in case someone needs it or wants to try it, for TX16S with internal GPS.
The default setting for EdgeTX is for GPSes that communicate at 9600 baud. Some GPS (false ublox) do not support changing the communication speed and being able to adjust it to the 9,600 bauds of EdgeTX.
I leave one for 38400 and one for 57600 baud.

`-PCB=X10 -PCBREV=TX16S -DEFAULT_MODE=2 GVARS=YES PPM_UNIT=US -INTERNAL_GPS_BAUDRATE:STRING=38400`

Erased 2.8 rc3 and added the final release

[EDGETX 2.8 GPS FIRM.zip](https://github.com/EdgeTX/edgetx/files/10118553/EDGETX.2.8.GPS.FIRM.zip)

**Comment by Taes59** [comment]
Buenas tardes, podriais hacer la compilacion en español. Muchas gracias.

**Reply by pfeerick** [reply]
It should already be available from https://github.com/pfeerick/lang-firmwares/releases/tag/release as per the release notes - choose the zip file for your transmitter, and look for the `ES` firmware. Next time, please remember to specify your transmitter also.

**Comment by Taes59** [comment]
Muchas gracias.

**Comment by sidewayz-control** [comment]
Could I have a standard firmware build for Radiomaster TX16s (Mk1) but with bluetooth enable. I have retrofitted the FrSky PARA bluetooth module to the mainboard.

Thanks

**Reply by tedsears** [reply]
I would also love this when you get a moment, same situation...

**Reply by pfeerick** [reply]
Done. Check the "By Request" link in the first post. 
```
cmake -DPCB=X10 -DPCBREV=TX16S -DBLUETOOTH=Y ..
```

**Reply by ArdSchell** [reply]
Thanks for the build of the TX16S with Bleutooth.
Can this already be turned ourself in Companion 2.8 or will this be posible in the next version?

**Comment by sande005** [comment]
Would like a standard build for a FrSky X10S Express, with ppmus, noheli, and lua.  With the appropriate cmake, no issue making my own.  Just unclear as to what standard build options do not need to be stated, specing the right -DPCB for that Tx, etc...

**Reply by pfeerick** [reply]
I can build it tomorrow morning for you... but to figure out the cmake, for starters you can look at https://github.com/EdgeTX/edgetx/blob/main/tools/build-gh.sh to figure out the minimum needed cmake for any of the supported transmitters. In this case... 

```
-DPCB=X10 -DPCBREV=EXPRESS -DPXX1=YES
```

And for most of the options, you can look at https://github.com/EdgeTX/edgetx/wiki/Compilation-options (which is in need up updating, but for the most part is still correct). Lua has always been on by default. Noheli (or more correctly  - HELI=N) is now the default, so you're just after `PPM_UNIT=US`

Making the full cmake
```
cmake -Wno-dev -DPCB=X10 -DPCBREV=EXPRESS -DPXX1=YES -DPPM_UNIT=US ..
```

**Reply by sande005** [reply]
Thank you Peter - I'll have a go on building it myself!  If I was more than really basic in the skills, I'd even take a crack at a small utility - kind of like Companion used to do, that would allow the user to use picklists and  tic boxes to generate the custom firmware.  Likely that is somewhere on a future "to-do" list, though.  I appreciate the assistance.  Will let you know if I run into trouble.
Ted

**Reply by sande005** [reply]
Seems to have worked - at least the radio boots up.  Did get one error:
ms-vscode.cpptools extension is not found in Open VSX .gitpod [Ln 5, Col 7]

Do I care?

**Reply by sande005** [reply]
Oops - for some reason the min/max for existing models still is in terms of %.  Here's the cmake I used:
cmake -Wno-dev -DPCB=X10 -DPCBREV=EXPRESS -DDEFAULT_MODE=2 -DGVARS=YES -DPPM_UNIT=US -DCMAKE_BUILD_TYPE=Release ../

**Reply by pfeerick** [reply]
I have this sneaky suspicion that this is because it's not been fully implemented. Can you remind me in a couple of days' time to have a look at this if I don't check back in?

**Reply by sande005** [reply]
Have you had a chance to look at this?  I'm away from my radio for a couple of months, so my need is not very critical.  Should I just open a regular GitHub case as a reminder?

**Comment by rydfree** [comment]
Hello , I am searching for a 2.8 build with the Heli option. I have been searching but may have over looked something. I have just purchased a Radiomaster T16 MkII and have 5 flybarred helis that I would like to set up . I currently fly them with my old Jumper T16 but would like to consolidate all my aircraft into the one radio. Any help would be appreciated.

**Reply by pfeerick** [reply]
You'll be able to get a heli build from https://github.com/pfeerick/lang-firmwares/releases/tag/latest - you want the [tx16s-heli-lang-e6c8e7c.zip](https://github.com/pfeerick/lang-firmwares/releases/download/latest/tx16s-heli-lang-e6c8e7c.zip) for the TX16S with HELI enabled, and then it's just a matter of picking the your desired language from that file.

**Reply by rydfree** [reply]
That was fast . Thanks a bunch

**Comment by xkam1x** [comment]
Hello Team,

I have an unusual request so please stay with me. I have a Taranis X9D Plus which support ACCST with an internal XJT module. This transmitter is pre 2019 version and its equipped with STM32F2 processor. I also use Lua telemetry, particularly yapoo FrSky and kept running into 'script error low memory'. Unfortunately once the script crashes and the error appears, rebooting controller is needed but not possible if drone is in air.

I found that STM32 family of MCUs are designed in such a way that make it possible for developer to swap MCUs with high tier counterpart. That is exactly what I have done!. I have replaced STM32F205 with STM32F407. Currently it's running the same original firmware which means that new MCU is running under clocked.

I would to know if its possible to have firmware that is able to take full advantage of new faster MCU in older board. Or would like to know if the additional RAM is being utilised. (I think it is as before I could only run one script and now have yapoo, numbers and bars running simultaneously)

Best Wishes,
Kam

psst: I am also working on going further and upgrading to STM32F427 with 256k ram.

**Reply by pfeerick** [reply]
This is probably best asked on the EdgeTX discord. Anything is possible given this is open source software, but obviously, changes would need to be made to suit the new MCU... if you use an MCU that is supported by another radio, it may be relatively trivial to port it over.

**Reply by rotorman** [reply]
@xkam1x Is the X'tal on your X9D+ 12 MHz?
If so, as X9D+2019 uses STM32F407
https://github.com/EdgeTX/edgetx/blob/df00d7d8a02e57497a16ccc54e298f025fd41c8a/radio/src/targets/taranis/CMakeLists.txt#L70
you could use this target. But you need to alter some GPIO pin mapping in https://github.com/EdgeTX/edgetx/blob/main/radio/src/targets/taranis/hal.h to match your X9D+, as there are some changes, e.g. https://github.com/EdgeTX/edgetx/blob/df00d7d8a02e57497a16ccc54e298f025fd41c8a/radio/src/targets/taranis/hal.h#L1209-L1214
Best, you make yourself a pinmapping table by scanning the hal.h and listing full microcontroller I/O mapping for X9DP and X9DP2019 and then compare them.

**Reply by xkam1x** [reply]
@pfeerick @rotorman I have achieved something:
I checked the oscillator with oscilloscope and it was 12MHz. looked at hal.h for pre 2019 and 2019 boards and found they are not compatible, GPIOs have moved for the same physical function probably to aid updated PCB layout and some functions have changed.
Looking at F205 and F407 data sheet, they are almost pin2pin compatible and have mostly the same internal peripherals.
I then forked edgeTX main and found `radio/src/targets/taranis/CMakeLists.txt` where I found PCB X9D+s CPU configuration:
Original
```
set(CPU_TYPE STM32F2)
set(CPU_TYPE_FULL STM32F205xE)  # for size report
add_definitions(-DSTM32F205xx)
set(LINKER_SCRIPT targets/taranis/stm32f2_flash.ld)
```
and changed them to:
```
set(CPU_TYPE STM32F4)
set(CPU_TYPE_FULL STM32F407xE)  # for size report
add_definitions(-DSTM32F40_41xxx -DSTM32F407xx)
set(LINKER_SCRIPT targets/taranis/stm32f4_flash.ld)
```
This meant the pin mapping remained the same but CPU config now reports STM32F4 which is exactly the hardware I now have.
I compiled on `gitpod.io` with
```
cmake -Wno-dev -DPCB=X9D+ -DPCBREV=X9DP -DDEFAULT_MODE=2 -DGVARS=YES -DPPM_UNIT=US -DCMAKE_BUILD_TYPE=Release ../
```
```
make configure
```
```
make -C arm-none-eabi firmware
```
flashed the radio with generated firmware.bin and its alive. This wouldn't have been possible without open source software!
I don't have full understanding of edgeTX project to say that these are only changes necessary. I understand that every time a new version of edgeTX is released, I will have to make these changes to source and compile a firmware for my radio since it's the only one in existence with X9D+ PCB containing STM32F407 MCU.
How much work is needed to add support for new hardware and if you have a document that outlines the process that would be brilliant?

**Reply by rotorman** [reply]
There is work in progress to support unified hardware targets, which in short would mean that you will have a single file configuring the radio hardware. With this system in place in the future, you would build firmware for your radio with your custom hardware configuration file.

**Comment by C-mos80** [comment]
Hello. I have seen that on the By request folder there is already a tx16s file with bluetooth enabled. I tried it on mine but the bluetooth wont show an address. Could it be beacause i dont have the para bluetooth? I have the cheaper one from Frsky. If so, can i have a 2.8 build with bluetooth and noheli for the cheaper one?

Thanks in advance

**Reply by pfeerick** [reply]
AFAIK, there is only one 'built in' bluetooth support atm, so I don't think I can help there. 

@rotorman It is for the frsky para module, isn't it? Is Cliff's Bluetooth project far along enough yet to be used as a drop in replacement?

**Reply by rotorman** [reply]
Building firmware with -DBLUETOOTH=YES enables TX16S to support FrSky Bluetooth modules - both, older FrSky CC2540 as well as newer FrSky PARA Bluetooth modules (they use the same AT commands, some listed in https://www.rcgroups.com/forums/showpost.php?p=46878713&postcount=9918 ).
@C-mos80 Can you show a picture of which module you have and how you soldered/connected it.

ESP Bluetooth Wifi Module by @dlktdr (https://github.com/dlktdr/BTWifiModule and for TX16S also see https://btwifimod.gitbook.io/untitled/getting-started/wiring#example-setup-mounted-internally-to-a-radio-master-tx16s ) can be used for Bluetooth trainer, but not yet for Bluetooth telemetry.

**Comment by C-mos80** [comment]
![IMG_20230206_230622](https://user-images.githubusercontent.com/124683585/217484583-f0ba740b-5463-48b9-b519-a803a738ffb3.jpg)

To tell you the truth a friend of mine did the whole process.

**Reply by rotorman** [reply]
Is that module really from FrSky (or is it RF-Star RF-BM-S02 https://www.rfstariot.com/uploadfile/downloads/20200819%20RF-BM-S02I%20Hardware%20Datasheet%20V1.0.pdf and possibly lacking the correct CC2540 firmware )?

**Reply by C-mos80** [reply]
It is not original Frsky. So could it be a bad Bluetooth module? I will try to buy a genuine from Frsky just to be sure. What i want is to control a headtracker via Bluetooth. Is it ok if i buy the cheap one or should i go for the para one?
![Screenshot_2023-02-08-13-24-47-380_com alibaba aliexpresshd](https://user-images.githubusercontent.com/124683585/217517541-455d3ebe-b0fa-4e54-a271-445979e0a0c4.jpg)

**Reply by rotorman** [reply]
Both should be OK. If you only want trainer capability, then getting ESP32, like this ESP32 C3 mini https://btwifimod.gitbook.io/untitled/getting-started/hardware would also work and be more future proof (search for ESP32-C3-DEVKITM-1 with integrated antenna or ESP32-C3-DEVKITM-1U for use with external antenna).

**Comment by OG-ALviK** [comment]
Is there a way to make a custom FW for the NV14 to use USART3 (which is currently hooked up to the unisable BLE module) as AUX_Serial like on the tx16s. So that it's possible to actually use it for various things like Serial Trainer, FrSky bluetooth modules or GPS modules. 
Either way that internal BLE module in the NV14 is just dead weight, might aswell flip it for something useful...  😂😂
